docs/dev/design/file-tree-diff.rst

```diff
@@ -0,0 +1,234 @@
+Version file tree diff
+======================
+
+Goals
+-----
+
+- Compare files from two versions to identify the files that have been added, removed, or modified.
+- Provide an API for this feature.
+- Integrate this feature to suggest redirects on files that were removed.
+- Integrate this feature to list the files that changed in a pull request.
+
+Non-goals
+---------
+
+- Replace the `docdiff <https://github.com/readthedocs/addons?tab=readme-ov-file#docdiff>`__ feature from addons.
+  That works on the client side, and it's good for comparing the content of files.
+
+Current problems
+----------------
+
+Currently, when a user opens a PRs, they need to manually search for the files of interest (new and modified files).
+We have a GitHub action that links to the root of the documentation preview, that helps a little, but it's not enough.
+
+When files are removed or renamed, users may not be aware that a redirect may be needed.
+We track 404s in our traffic analytics, but they don't keep track of the version,
+and it may be too late to add a redirect when users are already seeing a 404.
+
+In the past, we haven't implemented those features, because it's hard to map the source files to the generated files,
+since that depends on the build tool and configuration used by the project.
+
+Git providers may already offer a way to compare file trees, but again,
+they work on the source files, and not on the generated files.
+
+All hope was lost for having nice features like this, until now.
+
+Proposed solution
+-----------------
+
+Since redirects and files of interest are related to the generated files.
+Instead of working over the source files, we will work over the generated files, which we have access to.
+
+The key points of this feature are:
+
+- Get the diff of the file tree between two versions.
+- Expose that as an API.
+- Integrate that in PR previews.
+
+Diff between two S3 directories
+-------------------------------
+
+We are already using ``rclone`` to speed up uploads to S3,
+``rclone`` has a command (``rclone check``) to return the diff between two directories.
+For this, it uses the metadata of the files, like size and hash
+(it doesn't download the files).
+
+.. code:: console
+
+   $ ls a
+   changed.txt  new.txt  unchanged.txt
+   $ ls b
+   changed.txt  deleted.txt  unchanged.txt
+   $ rclone check --combined=- /usr/src/app/checkouts/readthedocs.org/a /usr/src/app/checkouts/readthedocs.org/b
+   + new.txt
+   - deleted.txt
+   = unchanged.txt
```

Comment:
Didn't see an option to exclude the unchanged files from the list, we could log each type to three different files if we want, but I like no having to deal with files...

Reply:
Yea, that would be nice, especially for very large projects since most files won't have changed. Could always do something like `grep -v "= "` or similar, but that still returns it, but we can avoid it in the Python process at least.